File: problems/Q3/Q3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        # charmap stores the current index of a character
        charmap = {}

        ans = int(n > 0) # if len is larger than 0, the output should be at least 1
        i = 0
        # try to extend the range [i, j]
        for j in range(n):
            i = max(charmap.get(s[j], -1), i) # search for nearest repeated char
            ans = max(ans, j - i + 1)   # calculate current length
            charmap[s[j]] = j + 1
        return ans

print(Solution().lengthOfLongestSubstring("au"))

File: problems/Q3/Q3.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	charmap := make(map[rune]int)

	res := 0
	i := 0

	for j, ch := range s {
		if value, exists := charmap[ch]; exists && i < value {
			i = value
		}
		if res < j-i+1 {
			res = j - i + 1
		}
		charmap[ch] = j + 1
	}

	return res
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabc"))
}
